sdk/core/azure-core/inc/azure/core/etag.hpp

```diff
@@ -0,0 +1,196 @@
+    {
+      // ETags are != if one of the values is null
+      if (!left.m_value || !right.m_value)
+      {
+        // Caveat, If both values are null then we consider the ETag equal
+        return !left.m_value && !right.m_value;
+      }
+
+      switch (comparisonKind)
+      {
+        case ETagComparison::Strong:
+          // Strong comparison
+          // If either is weak then there is no match
+          //  else tags must match character for character
+          return !left.IsWeak() && !right.IsWeak()
+              && (left.m_value.GetValue().compare(right.m_value.GetValue()) == 0);
+          break;
+
+        case ETagComparison::Weak:
+
+          if (left.IsWeak() && right.IsWeak())
+          {
+            return left.m_value.GetValue().compare(right.m_value.GetValue()) == 0;
+          }
+          auto leftStart = left.IsWeak() ? 2 : 0;
+          auto rightStart = right.IsWeak() ? 2 : 0;
+
+          auto leftVal = left.m_value.GetValue();
+          auto rightVal = right.m_value.GetValue();
+
+          // Compare is lengths are equal
+          return ((leftVal.length() - leftStart) == (rightVal.length() - rightStart))
+              && (leftVal.compare(leftStart, leftVal.length() - leftStart, &rightVal[rightStart])
+                  == 0);
+          break;
+      }
+      // Unknown comparison
+      abort();
+    }
+
+  protected:
+    /**
+     * @brief Construct an empty (null) #ETag.
+     */
+    ETag() = default;
+
+  public:
+    /**
+     * @brief Construct a #ETag.
+     */
+    explicit ETag(std::string etag) : m_value(std::move(etag)) {}
+
+    /**
+     * @brief Copy assignment.
+     */
+    ETag& operator=(const ETag&) = default;
+
+    /**
+     * @brief Whether @Tag is present.
+     * @return `true` if @ETag has a value, `false` otherwise.
+     */
+    bool HasValue() const { return m_value.HasValue(); }
+
+    /*
+     * @brief Returns the resource metadata represented as a string
+     * @return #std::string
+     */
+    const std::string& ToString() const
+    {
+      if (!m_value.HasValue())
+      {
+        abort();
+      }
+      return m_value.GetValue();
+    }
+
+    /**
+     * @brief Compare with \p other @ETag for equality.
+     * @param other Other @ETag to compare with.
+     * @return `true` if @ETag instances are equal, `false` otherwise.
+     */
+    bool operator==(const ETag& other) const
+    {
+      return Compare(*this, other, ETagComparison::Strong);
+    }
+
+    /**
+     * @brief Compare with \p other @ETag for inequality.
+     * @param other Other @ETag to compare with.
+     * @return `true` if @ETag instances are not equal, `false` otherwise.
+     */
+    bool operator!=(const ETag& other) const { return !(*this == other); }
+
+    /**
+     * @brief Specifies whether the @ETag is strong or weak
+     * @return `true` if @ETag is a weak validator, `false` otherwise.
+     */
+    bool IsWeak() const
```

Comment:
We could add it later if needed.  Its rare that customers will use the IsWeak() and doing !IsWeak() seems reasonable

Reply:
I don't think weak etag is used anywhere. So it's totally fine if we just support strong etag. We can add weak etag later if needed (I assume it will never be needed).